### PR TITLE
fix #17072: add times.dateTime with ISO 8601 order

### LINKED ISSUE #17072: deprecate `initDateTime` which has has unintuitive parameter order, introduce new set of inits
https://github.com/nim-lang/Nim/blob/f49d148a918cf8becec91f7d06f9b72e2bf4c4a2/lib/pure/times.nim#L1326

It turns out that `initDateTime` expects day of month first, which is of course a local convention in some regions of the world, but quite.. unusual in engineering - https://www.iso.org/iso-8601-date-and-time-format.html is generally used to avoid ambiguity, retain trivial lexicographic sorting and to maintain sanity in general when switching from pretty much any language out there where year-month-date really is an established canonical order for these parameters - also, thinking about it from a logical simplicity point of view, it's way easier to remember that it goes from least granular to most granular: `yyyy-mm-dd hh:mm:ss` - switching order in the middle of the parameters is quite confusing.

The request here is to deprecate the existing functions and replacing them with a new set of functions that adhere to relevant standards, are logically consistent and aren't just unique for uniqueness' sake.

### PR DESCRIPTION
**Pull request by timotheecour**
fix #17072
closes #18584

## links
* https://en.wikipedia.org/wiki/ISO_8601
* https://www.iso.org/iso-8601-date-and-time-format.html

### CODE REVIEW

File: lib/pure/times.nim [hunk 1]
```diff
@@ -1333,10 +1333,30 @@ proc now*(): DateTime {.tags: [TimeEffect], benign.} =
   ##    `cpuTime` instead, depending on the use case.
   getTime().local
 
+proc dateTime*(year: int, month: Month, monthday: MonthdayRange, 
+                   hour: HourRange = 0, minute: MinuteRange = 0, second: SecondRange = 0,
+                   nanosecond: NanosecondRange = 0,
+                   zone: Timezone = local()): DateTime =
```
Comment (konsumlamm) [inline lib/pure/times.nim]:
```suggestion
proc dateTime*(year: int, month: Month, monthday: MonthdayRange, 
               hour: HourRange = 0, minute: MinuteRange = 0, second: SecondRange = 0,
               nanosecond: NanosecondRange = 0,
               zone: Timezone = local()): DateTime =
```

File: lib/pure/times.nim [hunk 2]
```diff
@@ -1333,10 +1333,30 @@ proc now*(): DateTime {.tags: [TimeEffect], benign.} =
...
+  let dt = DateTime(
+    monthdayZero: monthday,
+    year: year,
+    monthZero: month.int,
+    hour: hour,
+    minute: minute,
+    second: second,
+    nanosecond: nanosecond
+  )
+  result = initDateTime(zone.zonedTimeFromAdjTime(dt.toAdjTime), zone)
```
Comment (konsumlamm) [inline lib/pure/times.nim]: I'd make `initDateTime` call this proc, instead of duplicating the code.

File: lib/pure/times.nim [hunk 3]
```diff
@@ -1333,14 +1333,13 @@ proc now*(): DateTime {.tags: [TimeEffect], benign.} =
   ##    `cpuTime` instead, depending on the use case.
   getTime().local
 
-proc initDateTime*(monthday: MonthdayRange, month: Month, year: int,
-                   hour: HourRange, minute: MinuteRange, second: SecondRange,
-                   nanosecond: NanosecondRange,
-                   zone: Timezone = local()): DateTime =
+proc dateTime*(year: int, month: Month, monthday: MonthdayRange,
```
Comment (dom96) [inline lib/pure/times.nim:1336]: This naming goes against Nim's conventions :/

Comment (timotheecour) [reply lib/pure/times.nim:1336]: it's better than initDateTime2.

Comment (dom96) [reply lib/pure/times.nim:1336]: I guess, but maybe we should document this as being "allowed but discouraged" in our style guide

Comment (timotheecour) [reply lib/pure/times.nim:1336]: maybe; this can be addressed and discussed in a separate PR though

**Comment by Araq** [comment]
The examples showing `initDate` could use named parameters, no need for deprecation and a new constructor proc that does the same.

**Comment by timotheecour** [comment]
> The examples showing initDate could use named parameters, no need for deprecation and a new constructor proc that does the same.

Allowing one to write

```nim
dateTime(2017, mMar, 30)
```
instead of
```nim
initDateTime(year = 2017, month = mMar, monthday = 30, 00, 00, 00, 00, utc())
```
is a big usability improvement.

you upvoted https://github.com/nim-lang/Nim/issues/17072 btw

**Comment by Varriount** [comment]
Is there a per-module method to affect imports? For a change like this, where deprecation is *not* a clean answer (because deprecation doesn't free up identifiers!), a more elegant solution would be something like Python's `from __future__ import ...` mechanism. This would allow deprecating the old function *and* switching over to the new function, without having to resort to weird names.

Defines don't work, because they're program-wide.

Using import aliases works, somewhat, but requires the user to define the alias themselves.

**Comment by juancarlospaco** [comment]
> Is there a per-module method to affect imports?

`importutils` ?.

**Comment by Araq** [comment]
Here is my proposal:

1. No deprecation period. A switch `-d:nimPreviewInitDateTime` changes the parameter order. 
2. Code can support both old and new times.nim versions by using named parameters (and this should be documented well).
3. In 1.8 or 2.0 (whatever it is we release) `-d:nimPreviewInitDateTime` becomes the new default.

**Comment by timotheecour** [comment]
> Here is my proposal:

I don't like that proposal, you'd get 2 possibly interpretations (flipping monthday and year, both compatible with int) without any CT error nor warning, resulting in at best a RT exception, at worse (depending on -d:danger or year range used) a wrong runtime result.
```nim
var monthday = 30
var year = 2017
initDateTime(monthday, mMar, year, 00, 00, 00, 00, utc())
  # `RangeDefect` or silently wrong result depending on -d:danger

year = 28
initDateTime(monthday, mMar, year, 00, 00, 00, 00, utc()) # silently wrong results
```
```nim
import times
var a = 29
var b = 2020
echo initDateTime(a, mFeb, b, 00, 00, 00, utc()).isLeapDay # true
echo initDateTime(b, mFeb, a, 00, 00, 00, utc()).isLeapDay # false
```

And since you have no control over 3rd party dependencies, you'd be forced to use explicit named params everywhere, defeating purpose (and causing code to break when the switch is used contrary to semantics)

This PR is much simpler and safer.

**Comment by Araq** [comment]
> flipping monthday and year, both compatible with int

Ah, I thought you would always get a compile-time error if you get the order wrong... Too bad.

Maybe if we make year a distinct int type we can use overloading?

**Comment by timotheecour** [comment]
> Maybe if we make year a distinct int type we can use overloading?

this would work (unlike the previous proposal) but it's still less convenient than this PR; requiring `2017.Year` just for the year param looks just weird:
```nim
# added in times.nim
type Year* = distinct int
# use code:
echo initDateTime(2017.Year, mMar, 30)
echo initDateTime(2017.Year, mMar, 30, 10, 30, 45)
```

vs this PR:
```nim
echo dateTime(2017, mMar, 30)
echo dateTime(2017, mMar, 30, 10, 30, 45)
```

**Pull request merged**